### DISCUSSION TITLE
Websocket client heartbeat discarded in case of data received

### DISCUSSION THREAD

**Discussion by pohmelie** [aio-libs/aiohttp#8427] [Ideas]
According to code (and what I see as a side effect behaviour) client will reschedule `ping` call every time ws receive a message. In my case (I receive many messages and send nothing) this leads to server drop my connection as a silent one. I think ping/pong should not relly on activity and be simple as possible. Should I raise a bug on this?

https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client_ws.py#L261

**Comment by bdraco** [comment]
The `autoping` can be problem for high volume websocket traffic.  Changing the current implementation would likely be a breaking change for many use cases so If you need something custom, I'd recommend disabling `autoping` and scheduling them instead. 

Here is an example of a custom implementation https://github.com/home-assistant-libs/aioshelly/blob/main/aioshelly/rpc_device/wsrpc.py#L191

**Reply by pohmelie** [reply]
It is not an `autoping`, but `heartbeat` in terms of aiohttp api

**Comment by Dreamsorcerer** [comment]
> I think ping/pong should not rely on activity and be simple as possible.

The behaviour defined in the WebSocket spec is simply that upon receiving a PING frame you must respond with a PONG frame, nothing more. It is simply a convention that we use PING as a keepalive mechanism for the TCP connection (either for server side, and/or for the client to verify the server is still there).

> this leads to server drop my connection

As a server, we'd autoclose the connection after a timeout of not receiving a message. Receiving _any_ message from the client tells us that the connection is still alive and therefore the TCP keepalive should remain active. It's completely illogical for a server to close the connection when receiving messages, just because none of them are PING messages. So, this seems like faulty server design, rather than anything worth changing in aiohttp.

**Reply by pohmelie** [reply]
You get it wrong I think. I connect to ws, and receive text messages, not ping messages. And I want to send heartbeat (actually a ping) messages. aiohttp have such ability with `heartbeat=20` for example. But, since heartbeat routine cancelled on receive it never happens.

**Reply by Dreamsorcerer** [reply]
I think my response is still relevant. You're right that's it heartbeat, not autoping. autoping responds to PING from the server (and hides the messages from the receive() iterator, but does depend on messages being consumed in a timely manner), while heartbeat sends PING messages to the server.

I'd assume that the TCP ACK would be enough for the keepalive when the client receives a message. But, if not, maybe it makes more sense to do the heartbeat reset when sending messages, rather than receiving them.

**Comment by bdraco** [comment]
The ping code has been rewritten in the last 6 months. Please give 3.11.x a try
